### PR TITLE
Overhaul of CalculateContamination's model for determining hom alt sites

### PR DESCRIPTION
**Pull request by davidbenjamin**
@takutosato This dramatically improves `CalculateContamination` by giving more care to distinguishing hom alts from hets.  It makes an especially big difference in our tumor-only HCC1143 validations, where the accuracy is now very good (and BTW, ContEst gets these all completely wrong even *with* a matched normal).

It also makes the tool work better in targeted panels where there might not be enough hom alt sites by adding a backup hom ref mode that gets triggered automatically.  This is based on a Broadie request.

I will file a separate issue to update the docs.

### CODE REVIEW

**Comment by takutosato** [comment]
I'm on it

File: docs/mutect/mutect.tex [hunk 1]
```diff
@@ -350,15 +350,17 @@ \section{Calculating Contamination}
...
+
+We then perform maximum likelihood estimation MLE on a model with learned parameters $\mu$, the local minor allele fraction for each segment, $\chi$, the contamination, and a constant base error rate parameter $\epsilon$ determined by counting reads that are neither the ref nor primary alt base at biallelic SNPs as described above.  The model likelihood is
 \begin{equation}
-\prod_{{\rm hets~}s} \frac{1}{2} \left[ {\rm Binom}(a_s | d_s, f) + {\rm Binom}(a_s | d_s, 1 - f)    \right] \label{het-maf}
+P(\{ a\}| \{f\}, \chi, \{d\}    \prod_{{\rm segments~}n}\prod_{{\rm sites~}s} \sum_{{\rm genotype~}g} P(g|f_s) {\rm Binom}(a_s | d_s, (1 - \chi) \phi(g, \mu_n, \epsilon) + \chi f_s)
```
Comment (takutosato) [inline docs/mutect/mutect.tex]: closing paren and = missing after `P(\{ a\}| \{f\}, \chi, \{d\}`

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        }
+        return ImmutablePair.of(segments, mafs);
+    }
+
+    /**
+     * Calculate the contamination in a tumor sample using the hom alt and hom ref sites inferred from this model, which
+     * could be derived from the tumor itself or a matched normal.
+     * @return
+     */
+    public Pair<Double, Double> calculateContaminationFromHoms(List<PileupSummary> tumorSites) {
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java]: I think `tumorSites` can be final. Happens a few times in this file.

Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/contamination/ContaminationModel.java]: fixed

File: docs/mutect/mutect.tex (hunk 1 above)
Comment (davidbenjamin) [reply docs/mutect/mutect.tex]: Added a test for hom ref mode and caught/fixed a bug in the process.  Thanks!

**Comment by davidbenjamin** [comment]
Put in docs for the error bars.  Now this PR has no sloppy loose ends.

**Comment by davidbenjamin** [comment]
@takutosato back to you for another round of review.

File: docs/mutect/mutect.tex [hunk 3]
```diff
@@ -336,29 +336,31 @@ \section{Calculating Contamination}
...
+\begin{align}
+{\rm var}(XZ) =& E[X^2Z^2] - E[XZ]^2 \\
+=& (1 - f_s) E[X^2] - (1-f_s)^2 E[X]^2 \\
+=& (1 - f_s) \left( {\rm var}(X) + E[X]^2 \right) - (1 - f_s)^2E[X]^2 \\
+=& (1 - f_s) d_s \chi(1 - \chi) + f_s(1 - f_s) d_s^2 \chi^2
+\end{align}
+And therefore the standard error on $\hat{\chi}$ comes out to the square root of the sum of these per-site variances, divided by the denominator of Eq. Eq. \ref{contamination_estimate}, that is,
```
Comment (takutosato) [inline docs/mutect/mutect.tex]: Eq. Eq.

File: docs/mutect/mutect.tex [hunk 4]
```diff
@@ -336,29 +336,31 @@ \section{Calculating Contamination}
...
 
+Once we have learned the parameters of this model, we can easily infer the posterior probabilities of hom alt genotypes.  We take every site with a posterior probability greater than 0.5.  In order to make the result more reliable against CNVs, we again impose a threshold on segment minor allele fraction and apply the above formula only to hom alt sites in these segments.  This time, however, we choose the highest possible threshold such that the estimated relative error is less than 0.2.
 
+Finally, we note that the same calculation can be reverse by using alt reads in hom ref sites as the signal and replacing $f$ by $1-f$ everywhere above.  We use the estimate from hom refs as a backup when the hom alt estimate has too great an error, as can occur in the case of targeted panels with few sites.  We do not use this as our primary estimate because it is much more affected by uncertainty in the population allele frequencies and is thus susceptible to systematic bias.
```
Comment (takutosato) [inline docs/mutect/mutect.tex]: can be *reversed

File: docs/mutect/mutect.tex [hunk 5]
```diff
@@ -336,29 +336,31 @@ \section{Calculating Contamination}
...
...h priors $P({\rm hom~ref}) = (1-f_s)^2$, $P({\rm alt~minor}) = P({\rm alt~major}) = f_s(1-f_s)$, and $P({\rm hom alt} = f^2$.  $\phi(g, \mu, \epsilon)$ is the alt allele fraction of the uncontaminated sample: $\phi({\rm hom~ref}) = \epsilon$, $\phi({\rm alt~minor}) = \mu$, $\phi({\rm alt~major}) = 1 - \mu$, $\phi({\rm hom~alt}) = 1 - \epsilon$.  The binomial is the weighted average of the uncontaminated sample and sample reads drawn independently from allele frequency $f$.  This is inconsistent with a single diploid contaminant sample, or indeed with any finite number of contaminants, which is why we do not the the MLE estimate in the final output of the tool.  The model also assumes that the uncontaminated and contaminating samples have the same overall depth distribution at each site, which is inconsistent with any differences in copy-number.  We perform the MLE by brute force, alternately maximizing with respect to $\chi$ with $\mu$ fixed and vice versa.  In order to make the solution more robust, we exclude segments with low $\mu$ from the maximization over $\chi$ by taking the highest possible threshold (up to 0.5, of course) for $\mu$ that retains at least $1/4$ of all sites.
```
Comment (takutosato) [inline docs/mutect/mutect.tex]: which is why we do not *the the

File: docs/mutect/mutect.tex (hunk 3 above)
Comment (davidbenjamin) [reply docs/mutect/mutect.tex]: done

File: docs/mutect/mutect.tex (hunk 4 above)
Comment (davidbenjamin) [reply docs/mutect/mutect.tex]: fixed

File: docs/mutect/mutect.tex (hunk 5 above)
Comment (davidbenjamin) [reply docs/mutect/mutect.tex]: fixed

**Pull request merged**